> Joshua Bloch님의 Effective Java 3/E를 읽고 정리한 글입니다.
> 

## 명명 패턴보다 애너테이션을 사용하라

- JUnit 3과 같이 전통적으로 프로그램 요소에는 명명 패턴을 자주 적용했다.
- 명명 패턴을 사용하는 것보다 애너테이션을 사용하는 것이 왜 더 나은 방법인지 알아보자.

### 명명 패턴의 단점

- 오타가 난다면 제대로 동작하지 않는다.
    - 예) JUnit 3에서 실수로 테스트할 메서드 명을 `tsetXXX`로 한다면 동작을 하지 않는다.
- 올바른 프로그램 요소에서만 사용되리라 보증할 방법이 없다.
    - 예) 메서드가 아닌 클래스 이름을 TestXXX로 지어서 JUnit에게 전달하면 클래스명에는 테스트가 동작하지 않는다.
- 프로그램 요소를 매개변수로 전달할 방법이 없다.
    - 예) 특정 예외를 던져야 성공하는 테스트를 만들 수 없다.
    

### 애너테이션(Annotation)

- 코드 사이에 주석처럼 쓰여서 특별한 기능을 수행하도록 한다.

```java
/**
 * 테서트 메서드임을 선언하는 마커 애너테이션 예시
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface Test {
}
```

- 애너테이션 선언에 다는 애너테이션을 메타 애너테이션이라 한다.
- `@Retention` → 사용 용도에 따라 어노테이션 유지 범위를 설정하는 것
- `@Target` → 적용 대상을 지정한다. `ElementType.METHOD`의 경우 적용대상은 메서드다.
    - 애너테이션 적용 대상은 `java.lang.annotation.ElementType` 에 정의되어 있다.
- 아무 매개변수 없이 단순히 대상에 적용하는 애너테이션을 마커 애너테이션이라고 한다.

### 마커 애너테이션을 처리하는 프로그램 예시

```java
public class RunTests {

    public static void main(String[] args) throws ClassNotFoundException {
        int tests = 0;
        int passed = 0;
        Class<?> testClass = Class.forName(args[0]);
        for (Method m : testClass.getDeclaredMethods()) {
            if (m.isAnnotationPresent(Test.class)) {
                tests++;
                try {
                    m.invoke(null);
                    passed++;
                } catch (InvocationTargetException wrappedExc) {
                    Throwable exc = wrappedExc.getCause();
                    System.out.println(m + " 실패: " + exc);
                } catch (Exception e) {
                    System.out.println("잘못 사용한 @Test: " + m);
                }
            }
        }
        System.out.printf("성공: %d, 실패: %d%n", passed, tests - passed);
    }
}
```

- 해당 프로그램은 정규화된 클래스명을 받아 `@Test` 애너테이션이 적용된 메서드를 차례로 호출한다.
- 테스트 메서드가 예외를 던진다면 리플렉션 메커니즘이 `InvocationTargetException`로 감싸서 다시 던진다.
- `InvocationTargetException` 외의 예외가 발생했다면 `@Test` 애너테이션을 잘못 사용했다는 의미다.

### 매개변수 하나를 받는 애너테이션 타입

```java
/**
 * 명시한 예외를 던져야만 성공하는 테스트용 애너테이션
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface ExceptionTest {
    Class<? extends Throwable> value();
}
```

- `Throwable`을 확장한 클래스의 `Class` 객체를 가지고 있는 애너테이션이다.
- 다음은 실제 애너테이션의 적용과 활용 예시다.

```java
// 적용 예시
@ExceptionTest(ArithmeticException.class)
public static void m1() {
    int i = 0;
    i = i / i;
}

// 활용 예시 RunTests InvocationTargetException 예외 잡는 부분 수정
catch (InvocationTargetException wrappedExc) {
    Throwable exc = wrappedExc.getCause();
    Class<? extends Throwable> excType = m.getAnnotation(ExceptionTest.class).value();
    if (excType.isInstance(exc)) {
        passed++;
    }
}
```

- 해당 활용 예시는 애너테이션 매개변수의 값을 추출하여 테스트 메서드가 올바른 예외를 던지는지 확인한다.
- 발생 예외의 클래스 파일이 컴파일타임에는 존재했으나 런타임에는 존재하지 않는 경우 `TypeNotPresentException` 예외가 발생한다.

### 배열 매개변수를 받는 애너테이션 타입

```java
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface ArrayExceptionTest {
    Class<? extends Throwable>[] value();
}
```

- 전과 거의 동일하지만 배열 매개변수를 받는 애너테이션이다.
- 다음은 실제 애너테이션의 적용과 활용 예시다.

```java
// 적용 예시
@ArrayExceptionTest({ IndexOutOfBoundsException.class, NullPointerException.class})
public static void doublyBad() {
    List<String> list = new ArrayList<>();
    list.addAll(5, null);
}

// 활용 예시
catch (Throwable wrappedExc) {
    Throwable exc = wrappedExc.getCause();
    int oldPassed = passed;
    Class<? extends Throwable>[] excTypes = m.getAnnotation(ArrayExceptionTest.class).value();
    for (Class<? extends Throwable> excType : excTypes) {
        if (excType.isInstance(exc)) {
            passed++;
            break;
        }
    }
    if (passed == oldPassed)
        System.out.printf("TEST FAIL: %s", exc);
}
```

### 반복 가능한 애너테이션 타입

```java
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Repeatable(RepeatableExceptionTestContainer.class)
public @interface RepeatableExceptionTest {
    Class<? extends Throwable> value();
}

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
public @interface RepeatableExceptionTestContainer {
    RepeatableExceptionTest[] value();
}
```

- Java 8부터 여러 개의 값을 받은 애너테이션을 `@Repeatable` 메타애너테이션을 사용하여 만들 수 있다.
- @Repeatable을 단 애너테이션은 하나의 프로그램 요소에 여러 번 달 수 있다.
- 다음과 같은 주의점이 있다.
    - 컨테이너 애너테이션의 class 객체를 매개변수로 전달해야 한다.
    - 컨테이너 애너테이션은 내부 애너테이션 타입의 배열을 반환하는 `value()` 메서드를 정의해야 한다.
    - 컨테이너 애너테이션 타입에는 적절한 보존 정책과 적용 대상을 명시해야 한다.
- 다음은 실제 애너테이션의 적용과 활용 예시다.

```java
// 적용 예시
@RepeatableExceptionTest(IndexOutOfBoundsException.class)
@RepeatableExceptionTest(NullPointerException.class)
public static void doublyBad() {
    List<String> list = new ArrayList<>();
    list.addAll(5, null);
}

// 활용 예시
if (m.isAnnotationPresent(RepeatableExceptionTest.class)
        || m.isAnnotationPresent(RepeatableExceptionTestContainer.class)) {
    tests++;
    try {
        m.invoke(null);
        System.out.printf("TEST %s FAIL: 예외를 던지지 않음%n", m);
    } catch (Throwable wrappedExc) {
        Throwable exc = wrappedExc.getCause();
        int oldPassed = passed;
        RepeatableExceptionTest[] excTests = m.getAnnotationsByType(RepeatableExceptionTest.class);
        for (RepeatableExceptionTest excTest : excTests) {
            if (excTest.value().isInstance(exc)) {
                passed++;
                break;
            }
        }
        if (passed == oldPassed)
            System.out.printf("TEST FAIL: %s", exc);
    }
}
```

- `getAnnotationsByType()` 메서드는 반복 가능 애너테이션과 해당 컨테이너를 모두 가져온다.
- `isAnnotationPresent()` 의 경우 둘을 명확히 구분한다.
    - `@RepeatableExceptionTest` 단일 사용 시 → 컨테이너로 확인할 때 `false`
    - `@RepeatableExceptionTest` 복수 사용 시 → 반복 가능 애너테이션으로 확인할 때 `false`
- 반복 가능 애너테이션을 사용한다면 가독성을 개선할 수 있지만, 선언하고 처리하는 부분에는 코드가 복잡해진다.

## 정리

- 명명 패턴을 사용하기보단 자바가 제공하는 애너테이션 타입을 사용하자
- 애너테이션을 사용한다면 안전하고, 편리하고, 명명 패턴을 사용하는 것보다 다양한 기능을 제공할 수 있다.